### PR TITLE
Don't allow void to convert to object via user-defined conversion

### LINKED ISSUE #34872: Aspnetcore allowing void -> object casting
Original Issue: https://github.com/aspnet/AspNetCore/issues/9197

Here's a simple standalone repro on .NET Core 2.2:

```
using Microsoft.AspNetCore.Mvc;
using System;

namespace ConsoleApp2
{
    class Program
    {
        static void Main(string[] args)
        {
            new Test().FaultyMethod();
        }
    }

    public class Test
    {
        private void ReturnVoid() { }

        public ActionResult<object> FaultyMethod()
        {
            return ReturnVoid();
        }
    }
}


namespace Microsoft.AspNetCore.Mvc
{
    public sealed class ActionResult<TValue>
    {
        public ActionResult(TValue value)
        {
        }

        public static implicit operator ActionResult<TValue>(TValue value)
        {
            return new ActionResult<TValue>(value);
        }
    }
}
```

Run it and you get:

Unhandled Exception: System.InvalidProgramException: Common Language Runtime detected an invalid program.
   at ConsoleApp2.Test.FaultyMethod()
   at ConsoleApp2.Program.Main(String[] args) in C:\Users\elipton\source\repos\ConsoleApp2\ConsoleApp2\

### PR DESCRIPTION
**Pull request by RikkiGibson**
Fixes #34872

If anyone can think of any more cases that ensure we catch attempts to convert to/from void it would be appreciated.

### CODE REVIEW

**Comment by RikkiGibson** [comment]
@VSadov PTAL if you have the time

**Comment by jcouv** [comment]
@RikkiGibson There's a number of test failures in CI. Will hold off review until you refresh and ping again.

File: src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs [hunk 1]
```diff
@@ -182,6 +182,11 @@ public Conversion ClassifyConversionFromExpression(BoundExpression sourceExpress
             Debug.Assert(sourceExpression != null);
             Debug.Assert((object)destination != null);
 
+            if (sourceExpression.Type?.SpecialType == SpecialType.System_Void || destination.SpecialType == SpecialType.System_Void)
+            {
+                return Conversion.NoConversion;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs]: >return Conversion.NoConversion; [](start = 16, length = 31)

Do we need to add similar protection for other entry points in this class. For example, ```ClassifyConversionFromType``` below, etc.?

File: src/Compilers/CSharp/Test/Semantic/Semantics/UserDefinedConversionTests.cs [hunk 2]
```diff
@@ -1597,5 +1597,49 @@ static void Main()
 ";
             CreateEmptyCompilation(source).VerifyDiagnostics();
         }
+
+        [Fact, WorkItem(34876, "https://github.com/dotnet/roslyn/pull/34876")]
+        public void Repro_34876()
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/UserDefinedConversionTests.cs]: >Repro_34876 [](start = 20, length = 11)

Please add similar test for VB.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 2)

File: src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs (hunk 1 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs]: In principle I think yes. However I'm struggling to come up with a program that actually exercises a bad behavior around `void` conversion through a call to `ClassifyConversionFromType` or other public methods on Conversions.

File: src/Compilers/CSharp/Test/Semantic/Semantics/UserDefinedConversionTests.cs (hunk 2 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/UserDefinedConversionTests.cs]: Done

File: src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs [hunk 3]
```diff
@@ -168,6 +168,33 @@ public Conversion ClassifyConversionFromExpressionType(TypeSymbol source, TypeSy
...
+            // (e.g. for an expression lambda which returns void).
+            // Therefore we allow an identity conversion between 'void' and 'void'.
+            if (sourceIsVoid && destIsVoid)
+            {
+                conversion = Conversion.Identity;
+                return true;
+            }
+
+            // If exactly one of source or destination is of type 'void' then no conversion may exist.
+            if (sourceIsVoid || destIsVoid)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs]: >sourceIsVoid || destIsVoid [](start = 16, length = 26)

It looks like this method returns true for `source == null && destIsVoid`. Is there a test for that case?

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs]: I've played with a few scenarios such as the following:

```cs
class C
{
    void M(System.Action a) {}
    void M2()
    {
        M(() => null);
    }
}
```

where I suspect the compiler might check convertibility of an expression with no type to an expression with a void type. But I wasn't able to get the compiler to emit any diagnostics related to the failed conversion, or enter the `TryGetVoidConversion` method, or do anything else that felt clearly "testable".

File: src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs [hunk 4]
```diff
@@ -168,6 +168,33 @@ public Conversion ClassifyConversionFromExpressionType(TypeSymbol source, TypeSy
             return ClassifyConversionFromType(source, destination, ref useSiteDiagnostics);
         }
 
+        private bool TryGetVoidConversion(TypeSymbol source, TypeSymbol destination, out Conversion conversion)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs]: `static`

File: src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs [hunk 5]
```diff
@@ -168,6 +168,33 @@ public Conversion ClassifyConversionFromExpressionType(TypeSymbol source, TypeSy
...
+            if (sourceIsVoid || destIsVoid)
+            {
+                conversion = Conversion.NoConversion;
+                return true;
+            }
+
+            conversion = default;
+            return false;
+
+        }
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs]: Unnecessary blank line.

File: src/Compilers/CSharp/Test/Semantic/Semantics/UserDefinedConversionTests.cs (hunk 2 above)
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/UserDefinedConversionTests.cs]: Consider using the test name used in VB for these two tests.

**Pull request merged**